nx/lib/nx/binary_backend.ex

```diff
@@ -693,49 +693,48 @@ defmodule Nx.BinaryBackend do
     from_binary(out, data)
   end
 
-  defp element_add(_, a, b), do: a + b
-  defp element_subtract(_, a, b), do: a - b
-  defp element_multiply(_, a, b), do: a * b
-  defp element_divide(_, a, b), do: a / b
+  defp element_add(_, a, b), do: Complex.add(a, b)
```

Comment:
Putting convenience aside, I think it's confusing that `Complex.add(1, 1)` returns `2` instead of `%Complex{re: 2, im: 0}`. It is similar to how `Nx.add(1, 1)` returns a tensor and not `2`.

If we want to keep the current API, one option would be to document a canonical complex form, where numbers with imaginary part `0` are always returned as regular numbers, as in `Complex.add(~c(1, 2), ~c(-1, -2)) #=> 0`. Then it would be expected that `Complex.add(1, 1) #=> 2`. Not sure if that's the best idea though.

Alternatively, instead of matching for numeric arguments in `Complex`, we may as well have these matches here. I guess the main issue is that then we would need to duplicate the implementation of `erf`, etc?

Reply:
Third idea, we could extend the complex package, and define a `Number` module. Then `Number.add(a, b)` would support both built-in numbers and `Complex` numbers.

Reply:
I would rather have the Number module keeping this "no upcasts" behavior

Reply:
The problem with upcasting is that we cannot differ a complex with no imaginary part from a pure number after the result without checking the input types

Reply:
That's where canonicalization would help, if the imaginary part is 0 we would always downcast, regardless the input types. Then matching on regular number arguments in Complex would just be an optimisation.

But yeah, `Number` module seems like the most transparent option to me.

Reply:
Could also be `Number.Complex`, but this probably means more aliases (and I can imagine `Number` being taken by Elixir at some point). @josevalim will know best :)

Reply:
https://github.com/polvalente/elixir-complex/pull/4

@jonatanklosko @josevalim this is what Complex.Math would look like. Should we commit to this?

Reply:
Thanks for trying it out. Personally, I prefer `Complex`, without `Complex.Math`. If you always want a complex back, you can simply call `Complex.new()` before calling the other `Complex` functions.

Reply:
Theres always either ~z or ~i or ~j


z because it's normally what people use for complex variables, i and j due to the imaginary constant

Reply:
`~z` is kinda like `~x` for real numbers, but that said it seems the best to me, because it points to the number as a whole rather that the imaginary constant.